src/ucp/core/ucp_worker.c

```diff
@@ -2835,6 +2839,61 @@ static ucs_status_t ucp_worker_address_pack(ucp_worker_h worker,
                             address_length_p, (void**)address_p);
 }
 
+
+static size_t ucp_worker_get_max_am_header(ucp_worker_h worker)
+{
+    ucp_context_h context = worker->context;
+    uct_iface_attr_t *if_attr;
+    ucp_rsc_index_t iface_id;
+    size_t max_am_header, max_uct_fragment;
+    size_t max_rts_size, max_ucp_header;
+
+    if (worker->max_am_header != UCP_WORKER_MAX_AM_HEADER_UNINIT) {
+        goto ret;
+    }
+
+    if (!(context->config.features & UCP_FEATURE_AM)) {
+        worker->max_am_header = 0;
+        goto ret;
+    }
+
+    max_am_header  = SIZE_MAX;
+    max_rts_size   = sizeof(ucp_rndv_rts_hdr_t) +
+                     ucp_rkey_packed_size(context, UCS_MASK(context->num_mds),
+                                          UCS_SYS_DEVICE_ID_UNKNOWN, 0);
+    max_ucp_header = ucs_max(max_rts_size, UCP_AM_FIRST_FRAG_META_LEN);
+
+    /* Make sure maximal AM header can fit into one bcopy fragment
+     * together with RTS or first eager header (whatever is bigger)
+     */
+    for (iface_id = 0; iface_id < worker->num_ifaces; ++iface_id) {
+        if_attr = &worker->ifaces[iface_id]->attr;
+
+        /* UCT_IFACE_FLAG_AM_BCOPY is required by UCP AM feature, therefore
+         * at least one interface should support it.
+         * Make sure that except user header single UCT fragment can fit
+         * first fragment header and footer and at least 1 byte of data. It is
+         * needed to correctly use generic AM based multi-fragment protocols,
+         * which expect some amount of payload to be packed to the first
+         * fragment.
+         * TODO: fix generic AM based multi-fragment protocols, so that this
+         * trick is not needed.
+         */
+        if (if_attr->cap.flags & UCT_IFACE_FLAG_AM_BCOPY) {
+            max_uct_fragment = ucs_max(if_attr->cap.am.max_bcopy,
+                                       max_ucp_header - 1) - max_ucp_header - 1;
+            max_am_header    = ucs_min(max_am_header, max_uct_fragment);
+        }
+    }
+
+    ucs_assert(max_am_header < SIZE_MAX);
```

Comment:
this is not consistent now, because we use `INT32_MAX` as uninitialized value, so we implicitly assume that the size is less than `INT32_MAX`, right? Maybe use `SIZE_MAX` as uninitialized value instead?